src/mlpack/methods/gmm/eigenvalue_ratio_constraint.hpp

```diff
@@ -70,12 +70,33 @@ class EigenvalueRatioConstraint
     // shouldn't be any negative eigenvalues anyway, so it doesn't matter if we
     // are suddenly forcing them to be positive.  If the first eigenvalue is
     // negative, well, there are going to be some problems later...
-    eigenvalues = (eigenvalues[0] * ratios);
+    eigenvalues = eigenvalues.each_col() % ratios;
 
     // Reassemble the matrix.
     covariance = eigenvectors * arma::diagmat(eigenvalues) * eigenvectors.t();
   }
 
+  /**
+   * Apply the eigenvalue ratio constraint to the given diagonal covariance
+   * matrix (represented as a vector).
+   */
+  void ApplyConstraint(arma::vec& diagCovariance) const
+  {
+    // The matrix is already eigendecomposed but we need to sort the elements.
+    arma::uvec eigvalOrder = arma::sort_index(diagCovariance);
+    arma::vec eigvals = diagCovariance(eigvalOrder);
+
+    // Change the eigenvalues to what we are forcing them to be.  There
+    // shouldn't be any negative eigenvalues anyway, so it doesn't matter if we
+    // are suddenly forcing them to be positive.  If the first eigenvalue is
+    // negative, well, there are going to be some problems later...
+    eigvals = eigvals.each_col() % ratios;
```

Comment:
In the `EigenvalueRatioConstraint`, is it right the bug you said?

Reply:
The code I wrote doesn't have the bug but basically the eigenvalues are ordered backwards for the full covariance matrix implementation.

Reply:
When I tested it, `eigvals[0] * ratios` doesn't work.

Reply:
Well, I'll revert the change at line 73 in `eigenvalue_ratio_constraint.hpp`